### PR TITLE
Invite users to workspace

### PR DESCRIPTION
**Pull request by ShevchukTania**
**STATUS**
READY

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

image
image

### CODE REVIEW

File: app/controllers/account/workspaces_controller.rb [hunk 1]
```diff
@@ -38,10 +39,31 @@ def destroy
...
+    @workspace = resource
+    @invitation = @workspace.shared_workspaces.build(user_id: member_params[:user])
+    @invitation.save
+
+    respond_to :js
+  end
 
+  def delete_member
+    @workspace = resource
+    @member_id = resource.members.find(params[:user]).id
```
Comment (romaslmd) [inline app/controllers/account/workspaces_controller.rb]: he? what's going on?

File: app/controllers/account/workspaces_controller.rb [hunk 2]
```diff
@@ -38,10 +39,31 @@ def destroy
...
+    @result = resource.members.destroy(params[:user])
```
Comment (romaslmd) [inline app/controllers/account/workspaces_controller.rb]: fyi:
```
@user = @workplace.members.find(params[:user_id])
@user.destroy

respond_to :js
```

File: app/controllers/account/workspaces_controller.rb [hunk 3]
```diff
@@ -51,4 +73,8 @@ def resource
   def workspace_params
     params.require(:workspace).permit(:name).merge(user_id: current_user.id)
   end
+
+  def member_params
+    params.require(:workspace).permit(:user)
```
Comment (romaslmd) [inline app/controllers/account/workspaces_controller.rb]: something isn't right,  should be `:user_id`

File: app/helpers/account/workspaces_helper.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+module Account::WorkspacesHelper
+  def uninvited_users(workspace)
+    uninvited = User.all.select do |person|
```
Comment (romaslmd) [inline app/helpers/account/workspaces_helper.rb]: no way.. make this method better

File: app/views/account/workspaces/_new_member_link.html.slim [hunk 5]
```diff
@@ -0,0 +1,4 @@
+- if uninvited_users(@workspace).any?
```
Comment (romaslmd) [inline app/views/account/workspaces/_new_member_link.html.slim]: low performance

File: app/views/account/workspaces/delete_member.js.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+$('#workspace-member-<%= @member_id %>').remove();
```
Comment (romaslmd) [inline app/views/account/workspaces/delete_member.js.erb]: `@user.id`

**Comment by romaslmd** [comment]
no tests

File: app/controllers/account/workspaces_controller.rb [hunk 7]
```diff
@@ -5,6 +5,7 @@ def index
 
   def show
     @workspace = resource
+    @users = User.for_workspace(@workspace).order_desc
```
Comment (igorkasyanchuk) [inline app/controllers/account/workspaces_controller.rb]: rename variable to what it do, for example "@members"

File: app/controllers/account/workspaces_controller.rb [hunk 8]
```diff
@@ -38,10 +39,34 @@ def destroy
...
+      redirect_to account_workspaces_path, notice: 'You have removed yourself from workspace!'
+    else
+      respond_to :js
+    end
+  end
+
+  private
   def collection
-    current_user.workspaces
+    current_user.workspaces.union(current_user.invited_workspaces)
```
Comment (igorkasyanchuk) [inline app/controllers/account/workspaces_controller.rb]: add method in user#available_workspaces which returns `self.workspaces.union(self.invited_workspaces)` and call it here

File: app/helpers/account/workspaces_helper.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+module Account::WorkspacesHelper
+  def uninvited_users(workspace)
+    ids = workspace.members.ids << workspace.user.id
```
Comment (igorkasyanchuk) [inline app/helpers/account/workspaces_helper.rb]: I don't like such code
do in one row one operation

ids   = [workspace.user.id]
ids += workspace.members.ids
User.where.not(id: ids)

File: app/models/user.rb [hunk 10]
```diff
@@ -91,4 +94,8 @@ def watching?(task)
   def with_avatar?
     avatar&.attachment&.blob&.persisted?
   end
+
+  def self.for_workspace(workspace)
+    workspace.members.union(self.where(id: workspace.user.id))
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: workspace.user_id

File: app/views/account/workspaces/create_member.js.erb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<% if @invitation.persisted? %>
+  $('#workspace-members')
+  .append("<%=j render 'account/workspaces/show_user', users: [@invitation.user], workspace: @workspace %>")
+  $('#new-member').html("<%=j render 'account/workspaces/new_member_link', workspace: @workspace %>")
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/create_member.js.erb]: add space before j

File: app/views/account/workspaces/create_member.js.erb [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+<% else %>
+  $('#new-member').html("<%= j render 'member_form', workspace: @workspace %>");
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/create_member.js.erb]: full path to partial

File: app/views/account/workspaces/show.html.slim [hunk 13]
```diff
@@ -21,3 +21,8 @@
           .level-item
             = link_to "Delete",  account_workspace_project_path(@workspace, project),
               method: :delete, class: "button is-small right", data: { confirm: "Are you sure?" }
+    | Members:
+    span#new-member
+      = render 'new_member_link', workspace: @workspace
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/show.html.slim]: full path to partial

File: app/views/account/workspaces/show.html.slim [hunk 14]
```diff
@@ -21,3 +21,8 @@
...
+    #workspace-members
+      = render 'show_user', users: @users, workspace: @workspace
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/show.html.slim]: and here

File: config/routes.rb [hunk 15]
```diff
@@ -11,6 +11,11 @@
 
     resource :profile, only: [:edit, :update]
     resources :workspaces do
+      member do
```
Comment (igorkasyanchuk) [inline config/routes.rb]: it looks like could be extracted into new REST controller

Comment (igorkasyanchuk) [reply config/routes.rb]: create nested Members controller inside workspaces (member_id will be a user_id)

File: app/views/account/workspaces/_show_user.html.slim [hunk 16]
```diff
@@ -0,0 +1,9 @@
+- users.each do |member|
+  div data-attribute = "#{member.id}"
+    = fa_icon "user-o"
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/_show_user.html.slim]: fa_icon 'user-o', text: member.full_name you can do it like this and remove not needed span

File: app/views/account/workspaces/_member_form.html.slim [hunk 17]
```diff
@@ -0,0 +1,4 @@
+= simple_form_for :workspace, url: create_member_account_workspace_path(workspace), remote: true do |f|
+  = f.input :user_id, label: false, wrapper: :bulma_input, required: true, prompt: "Select user",
+                  collection: uninvited_users(workspace).map{ |user| [user.full_name, user.id] }
+  = f.submit 'Save', class: 'button is-link'
```
Comment (igorkasyanchuk) [inline app/views/account/workspaces/_member_form.html.slim]: rename button name to "Add member"

File: app/controllers/account/members_controller.rb [hunk 18]
```diff
@@ -0,0 +1,42 @@
+class Account::MembersController < Account::AccountController
+  def new
+    @workspace = resource
+    @shared_workspase = @workspace.shared_workspaces.build
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: @shared_workspaсe

File: app/controllers/account/workspaces_controller.rb [hunk 19]
```diff
@@ -5,6 +5,7 @@ def index
 
   def show
     @workspace = resource
+    @members = User.for_workspace(@workspace).order_desc
```
Comment (igorkasyanchuk) [inline app/controllers/account/workspaces_controller.rb]: @workspace.potential_members (implement a new method)
it's more readable and better place to return possible members

File: app/controllers/account/workspaces_controller.rb (hunk 8 above)
Comment (igorkasyanchuk) [reply app/controllers/account/workspaces_controller.rb]: available_workspaces? or not?

File: app/models/user.rb [hunk 20]
```diff
@@ -91,4 +94,12 @@ def watching?(task)
   def with_avatar?
     avatar&.attachment&.blob&.persisted?
   end
+
+  def self.for_workspace(workspace)
```
Comment (igorkasyanchuk) [inline app/models/user.rb]: move to @workspace.all_members to return all members of Workspace

File: app/views/account/members/_new_link.html.slim [hunk 21]
```diff
@@ -0,0 +1,3 @@
+- if uninvited_users(workspace).any?
```
Comment (igorkasyanchuk) [inline app/views/account/members/_new_link.html.slim]: - if workspace.potential_members.any?

**Review by igorkasyanchuk — APPROVED** [review]

**Pull request merged**